[Chore] Separate out hashing utilities from vllm.utils

dongbo910220 opened the pull request:
Part of #26900

This PR continues the effort to clean up `vllm.utils` by extracting hashing-related helper functions into a new, dedicated module `vllm/utils/hashing.py`. 

The following functions have been moved:

* vllm.utils.sha256 -> vllm.utils.hashing.sha256
* vllm.utils.sha256_cbor -> vllm.utils.hashing.sha256_cbor
* vllm.utils.get_hash_fn_by_name -> vllm.utils.hashing.get_hash_fn_by_name

CC: @DarkLight1337

### Test Plan


### Test Result

dongbo910220 committed:
refactor(utils): extract hashing helpers to dedicated module

- Move sha256, sha256_cbor, and get_hash_fn_by_name from vllm.utils.__init__
  to vllm/utils/hashing.py
- Update import in vllm/v1/engine/core.py to use new module path
- Add vllm/utils/hashing.py to pickle import allowlist
- Use direct 'import pickle' instead of importlib workaround

These hashing functions are internal-only, so no backward compatibility
is needed. This refactoring improves code organization while keeping
the implementation cleaner.

Signed-off-by: dongbo910220 <1275604947@qq.com>

DarkLight1337 commented on tools/pre_commit/check_pickle_imports.py:
```diff
@@ -43,6 +43,7 @@
     "tests/utils.py",
     # pickle and cloudpickle
     "vllm/utils/__init__.py",
+    "vllm/utils/hashing.py",
```
We should be able to remove `vllm/utils/__init__.py` from this list now

dongbo910220 replied:
Regarding removing `vllm/utils/__init__.py` from the allowlist: I kept it because of the
  `cloudpickle` usage in the `_run_method` function (line 1842). Would you prefer that I:
  1. Move `_run_method` to a separate module and add that to the allowlist instead?
  2. Refactor `_run_method` to avoid `cloudpickle`?
  3. Something else?

DarkLight1337 replied:
I see. Then no need to remove, thanks for explaining!

DarkLight1337 commented on vllm/utils/__init__.py:
```diff
@@ -64,6 +61,17 @@
+    sha256_cbor,
+)
+
+_ = (
```
No need to keep back-compatibility since it's supposed to only be used internally

DarkLight1337 replied:
Just make sure to update all vLLM code that uses these imports

dongbo910220 commented:
@DarkLight1337

  Thanks for the review! I've addressed all the feedback:

  **Changes made:**

  1. **Updated import paths**: Migrated all 7 files to use `from vllm.utils.hashing import ...`
  instead of importing from `vllm.utils`

  2. **Simplified pickle import**: Changed `hashing.py` to use direct `import pickle` and added it
   to the pre-commit allowlist

  3. **Removed backward compatibility**: Completely removed the re-exports from
  `vllm/utils/__init__.py` as suggested 

  4. **Allowlist note**: Kept `vllm/utils/__init__.py` in the allowlist since it still uses
  `cloudpickle` in `_run_method` (line 1842). Let me know if this should be handled differently.

  Verification:
  -  No remaining references to old import paths

  Let me know if anything else needs adjustment.
